stan/math/rev/mat/fun/cov_per.hpp

```diff
@@ -0,0 +1,347 @@
+#ifndef STAN_MATH_REV_MAT_FUN_COV_PER_HPP
+#define STAN_MATH_REV_MAT_FUN_COV_PER_HPP
+
+#include <stan/math/rev/core.hpp>
+#include <stan/math/rev/scal/fun/value_of.hpp>
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/scal/err/check_not_nan.hpp>
+#include <stan/math/prim/scal/err/check_positive.hpp>
+#include <stan/math/prim/scal/fun/square.hpp>
+#include <stan/math/prim/scal/fun/squared_distance.hpp>
+#include <stan/math/prim/scal/meta/scalar_type.hpp>
+#include <boost/math/tools/promotion.hpp>
+#include <boost/utility/enable_if.hpp>
+#include <boost/type_traits.hpp>
+#include <vector>
+#include <cmath>
+
+namespace stan {
+  namespace math {
+
+    /**
+     * This is a subclass of the vari class for precomputed
+     * gradients of cov_per.
+     *
+     * The class stores the double values for the distance
+     * matrix, pointers to the varis for the covariance
+     * matrix, along with a pointer to the vari for sigma,
+     * the vari for l and the vari for p.
+     *
+     * @tparam T_x type of std::vector of elements
+     * @tparam T_sigma type of sigma
+     * @tparam T_l type of length scale
+     * @tparam T_p type of period
+     */
+    template <typename T_x, typename T_sigma, typename T_l, typename T_p>
+    class cov_per_vari : public vari {
+    public:
+      const size_t size_;
+      const size_t size_ltri_;
+      const double l_d_;
+      const double sigma_d_;
+      const double p_d_;
+      const double sigma_sq_d_;
+      double* dist_;
+      double* sin_dist_;
+      double* cos_dist_;
+      double* sin_dist_sq_;
+      vari* l_vari_;
+      vari* sigma_vari_;
+      vari* p_vari_;
+      vari** cov_lower_;
+      vari** cov_diag_;
+
+      /**
+       * Constructor for cov_per.
```

Comment:
This could include the eqs. necessary for the autodiff. I don't think this is anything that is super required, but it could make someone elses' life easier (I had to go to the whiteboard and work things out myself to check).

Reply:
Added the formulas. Now I also saw that in the implementation it would be probably better to change 2sin(x)cos(x) to sin(2x)